### DISCUSSION TITLE
ROS support in yocto standard SDK

### DISCUSSION THREAD

**Discussion by jiaxshi** [bmwcarit/meta-ros#1204] [Ideas]
# Backgroud:
Build ROS package from scratch in yocto takes much time. You can use an SDK to independently develop and test code that is destined to run on some target machine.

There are several github issues related to SDK support.  And there's some work done by community members.
https://github.com/ros/meta-ros/pull/1203
https://github.com/ros/meta-ros/pull/988

I also did some work on kirkstone/humble, here are some issues/topics to be discussed.
#1215 

# Topics/issues
## How to add SDK support: packagegroup or inc file
- Packagegroup
Refer to https://github.com/ros/meta-ros/pull/1203. A new image recipe and packagegroup is added for SDK.

- inc file
I think SDK should be supported by default. Using a inc file and added in ros2.conf
```
diff --git a/meta-ros2/conf/distro/ros2.conf b/meta-ros2/conf/distro/ros2.conf
index 4fb656587..4178211a6 100644
--- a/meta-ros2/conf/distro/ros2.conf
+++ b/meta-ros2/conf/distro/ros2.conf
@@ -3,3 +3,4 @@
 DISTRO_NAME = "Robot Operating System 2 (ROS 2)"

 require conf/distro/include/ros-common.inc
+require conf/distro/include/ros2-sdk.inc
```

## "nativesdk" in bbappends
Refer to https://github.com/ros/meta-ros/pull/1203/commits/6670f76ad3d08687be2dad5869dc6fe366b74543, we need some bbappends for adding "nativesdk". 

Idea: a mechanism to add "nativesdk" in generated recipes.

### HOST path
- HOST path is included in cross compile
Similar to https://github.com/ros/meta-ros/pull/1192, if host path is included in xxx-export.cmake makes build failured.

- Setup script
No cross copile concept in python/colcon. So HOST path is using in artifacts of SDK output file
```
_colcon_prefix_chain_sh_COLCON_CURRENT_PREFIX=/local/mnt3/workspace/jiaxshi/projects/sdk/ws4/examples/install
```

### Python C library naming
For example:
```
install/pendulum_msgs/lib/python3.10/site-packages/pendulum_msgs/pendulum_msgs_s__rosidl_typesupport_fastrtps_c.cpython-310-x86_64-linux-gnu.so
```
expected naming:
```
install/pendulum_msgs/lib/python3.10/site-packages/pendulum_msgs/pendulum_msgs_s__rosidl_typesupport_fastrtps_c.cpython-310-aarch64-linux-gnu.so

```
- Temp solution
We can use cmake args to fix this issue.
```
-DPYTHON_SOABI=cpython-310-aarch64-linux-gnu
```

## How to install packages generated by SDK?  
ipk/deb packages are not generated by SDK.

How to run packages build by SDK?  Similar on Ubuntu platform？

**Comment by sgstreet** [comment]
### Using the SDK to build ROS2 packages

The ROS2 developers are used to sourcing the ROS2 `setup.bash` script and running `colcon`, something like this:

```
$ source /opt/ros/rolling/setup.bash
$ colcon build
```
To use the Yocto SDK you must source `environment-setup-cortexa76-oe-linux` script first, but the target sysroot `setup.bash` can not be used because it is not aware of the Yocto SDK structure.  Following enivronment is need to use the SDK:

```
export ROS_VERSION=2
export ROS_PYTHON_VERSION=3
export ROS_AUTOMATIC_DISCOVERY_RANGE=SUBNET
export ROS_DISTRO=rolling
export PYTHONPATH=$OECORE_NATIVE_SYSROOT/usr/lib/python3.12/site-packages:$OECORE_TARGET_SYSROOT/opt/ros/rolling/lib/python3.12/site-packages
export AMENT_PREFIX_PATH=$OECORE_TARGET_SYSROOT/opt/ros/rolling
usr/share:$OECORE_TARGET_SYSROOT/opt/ros/rolling/share:$OECORE_TARGET_SYSROOT/opt/ros/rolling/lib/cmake"
```
Should these exports somehow be added to the SDK `environment-setup-cortexa76-oe-linux` script or should a meta-ros SDK `setup.bash` be generated?

To get `colcon` to use the SDK the following incantation is required:

```
$ colcon build \
--merge-install \
--install-base $PWD/install/arm64 \
--cmake-args \
" -DCMAKE_TOOLCHAIN_FILE=$CMAKE_TOOLCHAIN_FILE" \
" -DCMAKE_STAGING_PREFIX=$PWD/install/arm64" \
" -DBUILD_TESTING=OFF"
```
Should we meta-ros patch `colcon` to include the required cmake-args by default?
